File: app/src/main/java/com/tapan/weather/forecast/core/network/callback/WeatherAPIHandler.java
package com.tapan.weather.forecast.core.network.callback;

import com.tapan.weather.forecast.core.network.request.WeatherRequest;
import com.tapan.weather.forecast.core.network.response.WeatherResponse;
import com.tapan.weather.forecast.core.network.service.WeatherService;
import io.reactivex.Single;

/**
 * API handler to fetch the weather content
 *
 * @author Tapan Rana (ttapan.rana@gmail.com)
 */
public class WeatherAPIHandler {

  private String baseUrl;

  public WeatherAPIHandler(String baseUrl) {
    this.baseUrl = baseUrl;
  }

  /**
   * Retrieve the content of weather details
   *
   * @return {@link Single} result of the content
   */
  public Single<WeatherResponse> getWeatherDetails(WeatherRequest request, String appId) {
    WeatherService weatherService = RetrofitHandler.getRetrofitInstance(baseUrl)
        .create(WeatherService.class);

    return weatherService.getWeatherDetails(request.getLatitude(), request.getLongitude(), appId);
  }
}

File: app/src/main/java/com/tapan/weather/forecast/core/network/callback/WeatherAPIHandler.kt
package com.tapan.weather.forecast.core.network.callback

import com.tapan.weather.forecast.core.network.request.WeatherRequest
import com.tapan.weather.forecast.core.network.response.WeatherResponse
import com.tapan.weather.forecast.core.network.service.WeatherService
import io.reactivex.Single

/**
 * API handler to fetch the weather content
 *
 * @author Tapan Rana (ttapan.rana@gmail.com)
 */
class WeatherAPIHandler(private val baseUrl: String) {

  /**
   * Retrieve the content of weather details
   *
   * @return [Single] result of the content
   */
  fun getWeatherDetails(
    request: WeatherRequest,
    appId: String
  ): Single<WeatherResponse> {
    val weatherService = RetrofitHandler.getRetrofitInstance(baseUrl)
        .create(WeatherService::class.java)
    return weatherService.getWeatherDetails(request.latitude, request.longitude, appId)
  }
}